supabase/migrations/17_ops_catalogs.sql

```diff
@@ -0,0 +1,128 @@
+-- Adds the ops_catalog_template and functions for publishing ops catalogs for each tenant.
+
+create table ops_catalog_template (
+	id flowid not null primary key check (id = '00:00:00:00:00:00:00:00'::flowid),
+
+	bundled_catalog json not null
+);
+
+comment on table ops_catalog_template is
+'This table contains a single value, which is the template that is used to publish ops catalogs for each tenant';
+
+comment on column ops_catalog_template.id is
+'The id is always the constant 00:00:00:00:00:00:00:00, which is just used to ensure that there is always at most one row';
+
+comment on column ops_catalog_template.bundled_catalog is
+'The bundled Flow catalog that will be used as a template for generating the ops catalog for each tenant.
+This will contain the placeholder TENANT, which will be replaced by the actual tenant name prior to publication.
+The value for this is generated by running `flowctl raw bundle` on the catalog specs in the `ops-catalog/` directory.
+Typically, this will actually get updated by using ops-catalog/generate-migration.sh';
+
+create function internal.create_ops_publication(tenant_prefix text, ops_user_id uuid)
+returns flowid as $$
+declare
+	new_draft_id flowid := internal.id_generator();
+	publication_id flowid := internal.id_generator();
+	ops_template json;
+	
+begin
+	select replace(bundled_catalog::text, 'TENANT', rtrim(tenant_prefix, '/'))::json into ops_template from ops_catalog_template where id =  '00:00:00:00:00:00:00:00';
+
+	insert into drafts (id, user_id, detail) values (new_draft_id, ops_user_id, 're-publishing ops catalog');
+	
+	-- Start by adding a deletion for all currently live ops catalog items.
+	insert into draft_specs (id, draft_id, catalog_name, spec_type, spec)
+		select
+			internal.id_generator() as id,
+			new_draft_id as draft_id,
+			catalog_name,
+			null as spec_type,
+			null as spec
+		from live_specs
+		where catalog_name like ('ops/' || tenant_prefix || '%');
+
+	insert into draft_specs (id, draft_id, catalog_name, spec_type, spec)
+		(
+		select
+			internal.id_generator() as id,
+			new_draft_id as draft_id,
+			"key" as catalog_name,
+			'capture'::catalog_spec_type as spec_type,
```

Comment:
is the `::catalog_spec_type` cast required, or just here as documentation? I'd think that would work implicitly.

Reply:
It was returning an error about mismatched types before I added the cast, but I never really dug into exactly why. I'd also expected it to work without the explicit cast :man_shrugging: